### PR TITLE
Add CMAKE_GENERATOR_PLATFORM

### LINKED ISSUE #21: activate.bat script defines non-existing Visual Studio 16 2019 Win64 CMake generator
- [x] I read [the conda-forge documentation](https://conda-forge.org/docs/user/introduction.html#how-can-i-install-packages-from-conda-forge) and could not find the solution for my problem there.

The PR https://github.com/conda-forge/vc-feedstock/pull/19 added support for using Visual Studio 16 2019 . In the `activate.bat` script, however, there was some code that set the [`CMAKE_GENERATOR`](https://cmake.org/cmake/help/latest/envvar/CMAKE_GENERATOR.html) env variable to the `Visual Studio 16 2019 Win64` value (as in the past it set it to `Visual Studio 15 2017 Win64`). However, while  `Visual Studio 15 2017 Win64` is a valid generator (see https://cmake.org/cmake/help/v3.19/generator/Visual%20Studio%2015%202017.html#platform-selection), the  `Visual Studio 16 2019 Win64` generator does not exists (see https://cmake.org/cmake/help/v3.19/generator/Visual%20Studio%2016%202019.html#platform-selection). 

The `Win64` suffix in the generator was removed as now the same objective can be obtained in more scalable way by setting the [`CMAKE_GENERATOR_PLATFORM`](https://cmake.org/cmake/help/latest/envvar/CMAKE_GENERATOR_PLATFORM.html#envvar:CMAKE_GENERATOR_PLATFORM) env variable to `x64`, and that should work with both VS2017 and VS2019.

### PR DESCRIPTION
**Pull request by h-vetinari**
Fixes #21

### CODE REVIEW

**Review by isuruf — CHANGES_REQUESTED** [review]
This introduces more problems though. If somebody does `cmake -G %CMAKE_GENERATOR%` in the recipe, `CMAKE_GENERATOR_PLATFORM` is ignored.

**Comment by h-vetinari** [comment]
> If somebody does `cmake -G %CMAKE_GENERATOR%`

Why would they do that (genuine question)? Since `CMAKE_GENERATOR` is set by the activation script (and automatically picked up by CMake), I don't see why that would make sense. Have you seen many recipes like that in the wild?

**Comment by traversaro** [comment]
> 
> 
> This introduces more problems though. If somebody does `cmake -G %CMAKE_GENERATOR%` in the recipe, `CMAKE_GENERATOR_PLATFORM` is ignored.

The problem is that since Visual Studio 2019 CMake stopped provide all the possible combinations as different generators (see https://cmake.org/cmake/help/latest/generator/Visual%20Studio%2016%202019.html#platform-selection vs https://cmake.org/cmake/help/latest/generator/Visual%20Studio%2015%202017.html#platform-selection) so you can't in any case influence the platform setting by just setting the generator.

File: recipe/activate.bat [hunk 1]
```diff
@@ -63,19 +63,22 @@ if errorlevel 1 (
...
 )
 
+
 IF "@{target}" == "win-64" (
-  set "CMAKE_GEN=Visual Studio @{ver} @{year} Win64"
+  set "CMAKE_PLAT=x64"
   set "BITS=64"
 ) else (
-  set "CMAKE_GEN=Visual Studio @{ver} @{year}"
+  set "CMAKE_PLAT=x86"
```
Comment (traversaro) [inline recipe/activate.bat:71]: The platform name for 32-bit Windows is Win32, see https://cmake.org/cmake/help/latest/generator/Visual%20Studio%2015%202017.html#platform-selection .
```suggestion
  set "CMAKE_PLAT=Win32"
```

**Comment by isuruf** [comment]
Sure, that's true for VS 2019, but this PR is changing VS 2017 too.

**Comment by traversaro** [comment]
Ah, I see, that's true.

> 
> 
> Sure, that's true for VS 2019, but this PR is changing VS 2017 too.

Ah, I got it then.

**Comment by h-vetinari** [comment]
I had been under the impression from @traversaro's OP in #21 that this is (CMake-)compatible across VS 2017/2019?

**Comment by h-vetinari** [comment]
> @isuruf: Sure, that's true for VS 2019, but this PR is changing VS 2017 too.

I also checked the x-compat in the docs [here](https://cmake.org/cmake/help/latest/generator/Visual%20Studio%2015%202017.html#platform-selection) [my emphasis]: 

> The CMAKE_GENERATOR_PLATFORM variable may be set, perhaps via the cmake(1) -A option, to specify a target platform name (architecture). For example:
>  *  cmake -G "Visual Studio 15 2017" -A Win32
>  *  cmake -G "Visual Studio 15 2017" -A x64
>  *  cmake -G "Visual Studio 15 2017" -A ARM
>  *  cmake -G "Visual Studio 15 2017" -A ARM64
>
> _**For compatibility with CMake versions prior to 3.1, one may specify a target platform name optionally at the end of the generator name. This is supported only for:**_
>Visual Studio 15 2017 Win64
> >   Specify target platform x64.
>
> Visual Studio 15 2017 ARM
> >  Specify target platform ARM.

**Comment by h-vetinari** [comment]
However, if you prefer to make `activate.bat` version dependent, that's not a problem for me either.

**Comment by gabm** [comment]
I have been bitten by the bug as well and wanted to work on a fix. Lucky I saw this PR.

Given that the change is backward compatible with vs2017, can we get this PR merged? If not, how can I help to improve?

**Comment by isuruf** [comment]
> Given that the change is backward compatible with vs2017

It's not due to reasons I mentioned. This change should be applied only for vs2019.

**Comment by traversaro** [comment]
To clarify a bit, I think that what @isuruf is suggesting is to have this logic (pseudocode is following, @isuruf feel free to correct me):
~~~
if(vs2017)
  if(32 bit)
    set CMAKE_GENERATOR env variable to "Visual Studio 15 2017"
  elseif(64 bit)
    set CMAKE_GENERATOR env variable to "Visual Studio 15 2017 Win64"
  endif()
elseif(vs2019)
  if(32 bit)
    set CMAKE_GENERATOR env variable to to "Visual Studio 16 2019"
    set CMAKE_PLATFORM env variable to to "Win32"
  elseif(64 bit)
    set CMAKE_GENERATOR env variable to to "Visual Studio 16 2019"
    set CMAKE_PLATFORM env variable to to "x64"
  endif()
endif()
~~~

**Comment by traversaro** [comment]
> 
> 
> To clarify a bit, I think that what @isuruf is suggesting for maximize backward compatibility is to have this logic (pseudocode is following, @isuruf feel free to correct me):
> 
> ```
> if(vs2017)
>   if(32 bit)
>     set CMAKE_GENERATOR env variable to "Visual Studio 15 2017"
>   elseif(64 bit)
>     set CMAKE_GENERATOR env variable to "Visual Studio 15 2017 Win64"
>   endif()
> elseif(vs2019)
>   if(32 bit)
>     set CMAKE_GENERATOR env variable to to "Visual Studio 16 2019"
>     set CMAKE_PLATFORM env variable to to "Win32"
>   elseif(64 bit)
>     set CMAKE_GENERATOR env variable to to "Visual Studio 16 2019"
>     set CMAKE_PLATFORM env variable to to "x64"
>   endif()
> endif()
> ```

Great, I did not see that this was implemented in https://github.com/conda-forge/vc-feedstock/pull/35 by @gabm .

**Comment by h-vetinari** [comment]
Closing as superseded by #35

**Comment by gabm** [comment]
yes, sorry - i should have mentioned that here...

**Pull request closed without merging**